grpcbp/server_middlewares.go

```diff
@@ -142,3 +146,113 @@ func StartSpanFromGRPCContext(ctx context.Context, name string) (context.Context
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (_ interface{}, err error) {
+		start := time.Now()
+
+		method := methodSlug(info.FullMethod)
+		activeRequestLabels := prometheus.Labels{
+			localServiceLabel: serviceSlug,
+			methodLabel:       method,
+		}
+		serverActiveRequests.With(activeRequestLabels).Inc()
+
+		defer func() {
+			success := strconv.FormatBool(err == nil)
+			status, _ := status.FromError(err)
+
+			latencyLabels := prometheus.Labels{
+				localServiceLabel: serviceSlug,
+				methodLabel:       method,
+				typeLabel:         unary,
+				successLabel:      success,
+			}
+			serverLatencyDistribution.With(latencyLabels).Observe(time.Since(start).Seconds())
+
+			rpcCountLabels := prometheus.Labels{
+				localServiceLabel: serviceSlug,
+				methodLabel:       method,
+				typeLabel:         unary,
+				successLabel:      success,
+				codeLabel:         status.Code().String(),
+			}
+			serverRPCRequestCounter.With(rpcCountLabels).Inc()
+			serverActiveRequests.With(activeRequestLabels).Dec()
+		}()
+
+		return handler(ctx, req)
+	}
+}
+
+// InjectPrometheusStreamServerInterceptor is a server middleware that tracks
+// Prometheus metrics.
+//
+// It emits the following metrics:
+//
+// * grpc_server_active_requests gauge with labels:
+//
+//   - grpc_service: the local service slug, serviceSlug arg
+//   - grpc_method: the name of the method called on the gRPC service
+//
+// * grpc_server_latency_seconds histogram and grpc_server_requests_total
+//   counter with labels:
+//
+//   - grpc_service and grpc_method
+//   - grpc_success: "true" if status is OK, "false" otherwise
+//   - grpc_type: type of request, i.e unary
+//   - grpc_code: the human-readable status code, e.g. OK, Internal, etc
+func InjectPrometheusStreamServerInterceptor(serviceSlug string) grpc.StreamServerInterceptor {
+	return func(srv interface{}, stream grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) (err error) {
+		start := time.Now()
+
+		method := methodSlug(info.FullMethod)
+		activeRequestLabels := prometheus.Labels{
+			localServiceLabel: serviceSlug,
+			methodLabel:       method,
+		}
+		serverActiveRequests.With(activeRequestLabels).Inc()
+
+		defer func() {
+			success := strconv.FormatBool(err == nil)
+			status, _ := status.FromError(err)
+
+			latencyLabels := prometheus.Labels{
+				localServiceLabel: serviceSlug,
+				methodLabel:       method,
+				typeLabel:         serverStream,
+				successLabel:      success,
+			}
+			serverLatencyDistribution.With(latencyLabels).Observe(time.Since(start).Seconds())
+
+			rpcCountLabels := prometheus.Labels{
+				localServiceLabel: serviceSlug,
+				methodLabel:       method,
+				typeLabel:         serverStream,
+				successLabel:      success,
+				codeLabel:         status.Code().String(),
+			}
+			serverRPCRequestCounter.With(rpcCountLabels).Inc()
+			serverActiveRequests.With(activeRequestLabels).Dec()
```

Comment:
same here, I don't think the request handling is done here.

Reply:
i've been testing this more to confirm and it does appear the request handling is done here when `handler()` is called. I've been reading the grpc code to confirm, but I was struggling finding the specific code. So in the meantime, I'm adding more test cases to fully convince myself.